packages/context/src/loader.ts

```diff
@@ -0,0 +1,252 @@
+  };
+}
+/**
+ * Get UI5 project
+ *
+ * @param projectRoot path to root of a project
+ * @param appRoot path to root of an app
+ * @param manifest path to manifest defined in under app root
+ * @param manifestDetails details of manifest
+ */
+export async function getUI5Project(
+  projectRoot: string,
+  appRoot: string,
+  manifest: Manifest,
+  manifestDetails: ManifestDetails
+): Promise<UI5Project | undefined> {
+  const app = await getApp(projectRoot, appRoot, manifest, manifestDetails, {
+    kind: "UI5",
+    type: "UI5",
+  });
+  if (!app) {
+    return;
+  }
+  return {
+    type: UI5_PROJECT_TYPE,
+    root: projectRoot,
+    app,
+  };
+}
+
+/**
+ * Get CAP or UI5 project
+ *
+ * @param documentPath path to a file i.e absolute/path/webapp/ext/main/Main.view.xml
+ */
+export async function getProject(
+  documentPath: string
+): Promise<Project | undefined> {
+  const projectRoot = await getProjectRoot(documentPath);
+  if (!projectRoot) {
+    return;
+  }
+  const cachedProject = cache.getProject(projectRoot);
+  if (cachedProject) {
+    return cachedProject;
+  }
+  const appRoot = await findAppRoot(documentPath);
+  if (!appRoot) {
+    return;
+  }
+  const manifestPath = await findManifestPath(documentPath);
+  if (!manifestPath) {
+    return;
+  }
+  const manifest = await getUI5Manifest(manifestPath);
+  if (!manifest) {
+    return;
+  }
+  const manifestDetails = await getManifestDetails(documentPath);
+  if (!manifestDetails) {
+    return;
+  }
+  const projectInfo = await getProjectInfo(projectRoot);
+  if (projectInfo.type === CAP_PROJECT_TYPE) {
+    const capProject = await getCAPProject(
+      projectRoot,
+      projectInfo,
+      appRoot,
+      manifest,
+      manifestDetails
+    );
+    if (capProject) {
+      cache.setProject(projectRoot, capProject);
+    }
+    return capProject;
+  }
+  const ui5Project = await getUI5Project(
+    projectRoot,
+    appRoot,
+    manifest,
+    manifestDetails
+  );
+  if (ui5Project) {
+    cache.setProject(projectRoot, ui5Project);
+  }
+  return ui5Project;
+}
+
+/**
+ * Get cap services
+ *
+ * @param projectRoot path to root of a project
+ */
+async function getCapServices(
+  projectRoot: string
+): Promise<Map<string, string>> {
+  const cachedCapServices = cache.getCapServices(projectRoot);
+  if (cachedCapServices) {
+    return cachedCapServices;
+  }
+  const services = new Map<string, string>();
+  try {
+    const data = await getCapModelAndServices(projectRoot);
+    const cds = await loadModuleFromProject<any>(projectRoot, "@sap/cds");
+    for (const service of data.services) {
+      const metadataContent = cds.compile.to.edmx(data.model, {
+        version: "v4",
+        service: service.name,
+        odataForeignKeys: true,
+        odataContainment: false,
+      });
+
+      services.set(service.urlPath, metadataContent);
+    }
+  } catch (error) {
+    console.log(error);
+    cache.deleteCapServices(projectRoot);
```

Comment:
Why you are deleting cache entry when it doesn't exist (in line 167 we already checked it)?
Why not to return rejected promise in case of exception? You are putting into the cache and return to consumer potentially incomplete services in case of exception.

Reply:
Thanks adapted.